examples/src/bin/stopwatch.rs

```diff
@@ -0,0 +1,198 @@
+        pub fn new() -> Self {
+            Self {
+                elapsed: Duration::zero(),
+                start_moments: Vec::new(),
+                pause_moments: Vec::new(),
+                paused: true, // stopped by default; start by explicitly calling `.resume()`
+            }
+        }
+
+        fn last_start(&self) -> DateTime<Local> {
+            self.start_moments[self.start_moments.len() - 1]
+        }
+        fn pause(&mut self) {
+            assert!(self.paused == false, "Already paused!");
+            let moment = Local::now();
+            self.pause_moments.push(moment);
+            self.elapsed = self.elapsed + (moment - self.last_start());
+            self.paused = true;
+        }
+        fn resume(&mut self) {
+            assert!(self.paused == true, "Already running!");
+            self.start_moments.push(Local::now());
+            self.paused = false;
+        }
+        fn pause_or_resume(&mut self) {
+            if self.paused {
+                self.resume();
+            } else {
+                self.pause();
+            }
+        }
+        /// Read the total time elapsed
+        fn read(&self) -> Duration {
+            if self.paused {
+                self.elapsed
+            } else {
+                self.elapsed + (Local::now() - self.last_start())
+            }
+        }
+    }
+
+    /// Separating the `StopWatch` 'core' and the `StopWatchView` improves reusability
+    /// and flexibility. The user may implement their own `View`s, i.e. layouts, based
+    /// on the same `StopWatch` logic.
+    pub struct StopWatchView {
+        stopwatch: StopWatch,
+        on_stop: Option<Rc<dyn Fn(&mut Cursive, &StopWatch)>>,
+    }
+
+    impl StopWatchView {
+        pub fn new() -> Self {
+            Self {
+                stopwatch: StopWatch::new(),
+                on_stop: None,
+            }
+        }
+
+        /// Sets a callback to be used when `<Enter>` is pressed.
+        ///
+        /// The elapsed time will be given to the callback.
+        ///
+        /// See also cursive::views::select_view::SelectView::set_on_submit
+        pub fn set_on_stop<F, R>(&mut self, cb: F)
+        where
+            F: 'static + Fn(&mut Cursive, &StopWatch) -> R,
+        {
+            self.on_stop = Some(Rc::new(move |s, t| {
+                cb(s, t);
+            }));
+        }
+
+        pub fn on_stop<F, R>(self, cb: F) -> Self
+        where
+            F: 'static + Fn(&mut Cursive, &StopWatch) -> R,
+        {
+            self.with(|s| s.set_on_stop(cb))
+        }
+
+        fn stop(&mut self) -> EventResult {
+            let stopwatch = &mut self.stopwatch;
+            if !stopwatch.paused {
+                stopwatch.pause();
+            }
+            let result = if self.on_stop.is_some() {
+                let cb = self.on_stop.clone().unwrap();
+                let stopwatch_data = self.stopwatch.clone(); // TODO: remove clone
```

Comment:
Almost done... Except in line 163 where I haven't figured out how not to clone

Reply:
Yeaah... the issue is that the callback will be called later, possibly after _removing this view_. So we can't depend on any data in the view itself; the callback must be self-supported.
The solution is to either clone the data as you are doing (making it standalone), or wrap it in a `Rc` shared by both the view and the callback.

Reply:
I just came up with another approach: use `mem::replace` to replace `self.stopwatch` with a new, zeroed `StopWatch`, while getting the ownership of the 'freshly' produced data, which can then passed to the closure.

Reply:
You can use [`std::mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) which does this automatically for "default" types.